File: dotnet/base/Mcma.Worker/ProviderCollection.cs
using System;
using Mcma.Client;
using Mcma.Context;
using Mcma.Logging;
using Mcma.Data;

namespace Mcma.Worker
{
    public class ProviderCollection
    {
        private readonly ILoggerProvider _loggerProvider;
        private readonly IResourceManagerProvider _resourceManagerProvider;
        private readonly IDocumentDatabaseTableProvider _dbTableProvider;
        private readonly IAuthProvider _authProvider;
        private readonly IContextVariableProvider _environmentVariableProvider;

        public ProviderCollection(ILoggerProvider loggerProvider,
                                  IResourceManagerProvider resourceManagerProvider,
                                  IDocumentDatabaseTableProvider dbTableProvider,
                                  IAuthProvider authProvider,
                                  IContextVariableProvider environmentVariableProvider = null)
        {
            _loggerProvider = loggerProvider;
            _resourceManagerProvider = resourceManagerProvider;
            _dbTableProvider = dbTableProvider;
            _authProvider = authProvider;
            _environmentVariableProvider = environmentVariableProvider ?? new EnvironmentVariableProvider();
        }

        public ILoggerProvider LoggerProvider
            => _loggerProvider ?? throw new Exception($"{nameof(LoggerProvider)} not available.");

        public IResourceManagerProvider ResourceManagerProvider
            => _resourceManagerProvider ?? throw new Exception($"{nameof(ResourceManagerProvider)} not available.");

        public IDocumentDatabaseTableProvider DbTableProvider
            => _dbTableProvider ?? throw new Exception($"{nameof(DbTableProvider)} not available.");

        public IContextVariableProvider EnvironmentVariableProvider
            => _environmentVariableProvider ?? throw new Exception($"{nameof(EnvironmentVariableProvider)} not available.");

        public IAuthProvider AuthProvider
            => _authProvider ?? throw new Exception($"{nameof(AuthProvider)} not available.");
    }
}

File: nodejs/base/worker/lib/provider-collection.ts
import { ContextVariableProvider, LoggerProvider, McmaException } from "@mcma/core";
import { AuthProvider, ResourceManagerProvider } from "@mcma/client";
import { DocumentDatabaseTableProvider } from "@mcma/data";

export interface Providers {
    authProvider?: AuthProvider;
    contextVariableProvider?: ContextVariableProvider;
    dbTableProvider?: DocumentDatabaseTableProvider;
    loggerProvider?: LoggerProvider;
    resourceManagerProvider?: ResourceManagerProvider;
}

export class ProviderCollection implements Providers {
    private readonly _authProvider: AuthProvider;
    private readonly _contextVariableProvider: ContextVariableProvider;
    private readonly _dbTableProvider: DocumentDatabaseTableProvider;
    private readonly _loggerProvider: LoggerProvider;
    private readonly _resourceManagerProvider: ResourceManagerProvider;

    constructor(providers: Providers) {
        this._authProvider = providers?.authProvider;
        this._contextVariableProvider = providers?.contextVariableProvider;
        this._dbTableProvider = providers?.dbTableProvider;
        this._loggerProvider = providers?.loggerProvider;
        this._resourceManagerProvider = providers?.resourceManagerProvider;
    }

    get authProvider() {
        if (!this._authProvider) {
            throw new McmaException("AuthProvider not available");
        }
        return this._authProvider;
    }

    get contextVariableProvider() {
        if (!this._contextVariableProvider) {
            throw new McmaException("ContextVariableProvider not available");
        }
        return this._contextVariableProvider;
    }

    get dbTableProvider() {
        if (!this._dbTableProvider) {
            throw new McmaException("DbTableProvider not available");
        }
        return this._dbTableProvider;
    }

    get loggerProvider() {
        if (!this._loggerProvider) {
            throw new McmaException("LoggerProvider not available");
        }
        return this._loggerProvider;
    }

    get resourceManagerProvider() {
        if (!this._resourceManagerProvider) {
            throw new McmaException("ResourceManagerProvider not available");
        }
        return this._resourceManagerProvider;
    }
}
